tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerTests.cs

```diff
@@ -266,6 +267,145 @@ public void Dispose_should_dispose_the_server()
             channel.Should().NotBeNull();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void GetChannel_should_update_topology_and_clear_connection_pool_on_network_error_or_timeout(
+            [Values("timeout", "network")] string errorType,
+            [Values(false, true)] bool async)
+        {
+            var serverId = new ServerId(_clusterId, _endPoint);
+            var connectionId = new ConnectionId(serverId);
+            Exception innerMostException;
+            switch (errorType)
+            {
+                case "timeout":
+                    innerMostException = new SocketException((int)SocketError.TimedOut);
+                    break;
+                case "network":
+                    innerMostException = new SocketException((int)SocketError.NetworkUnreachable);
+                    break;
+                default: throw new ArgumentException("Unknown error type.");
+            }
+
+            var openConnectionException = new MongoConnectionException(connectionId, "Oops", new IOException("Cry", innerMostException));
+            var mockConnection = new Mock<IConnectionHandle>();
+            mockConnection.Setup(c => c.Open(It.IsAny<CancellationToken>())).Throws(openConnectionException);
+            mockConnection.Setup(c => c.OpenAsync(It.IsAny<CancellationToken>())).ThrowsAsync(openConnectionException);
+            var mockConnectionPool = new Mock<IConnectionPool>();
+            mockConnectionPool.Setup(p => p.AcquireConnection(It.IsAny<CancellationToken>())).Returns(mockConnection.Object);
+            mockConnectionPool.Setup(p => p.AcquireConnectionAsync(It.IsAny<CancellationToken>())).ReturnsAsync(mockConnection.Object);
+            var mockConnectionPoolFactory = new Mock<IConnectionPoolFactory>();
+            mockConnectionPoolFactory
+                .Setup(f => f.CreateConnectionPool(It.IsAny<ServerId>(), _endPoint))
+                .Returns(mockConnectionPool.Object);
+            var mockMonitorServerDescription = new ServerDescription(serverId, _endPoint);
+            var mockServerMonitor = new Mock<IServerMonitor>();
+            mockServerMonitor.SetupGet(m => m.Description).Returns(mockMonitorServerDescription);
+            void MockMonitorInvalidate(string reason)
```

Comment:
we usually place local methods at the end of method

Reply:
I think we were undecided the last time we visited this (with the counter argument being "why should the declaration order of locals methods and Actions/Funcs be different"), but in the interest of speed, we can discuss this style question later :) Moved!